File: Sources/MurmurDictionary/DictionaryCorrector.swift
import Foundation

/// One correction that actually fired, kept so history can show whether the dictionary is
/// earning its place.
public struct AppliedCorrection: Codable, Hashable, Sendable {
    /// Which kind of rule fired, so history can tell a repair from an expansion.
    public enum Kind: String, Codable, Sendable {
        case correction
        case snippet
        /// A word matched against something visible on screen — a filename, a symbol.
        /// Produced by the macOS app only; the shared correction pass never emits it.
        case screen
    }

    /// The text as the engine produced it.
    public let from: String
    /// What it was rewritten to.
    public let to: String
    /// How many times it fired in this transcript.
    public let count: Int
    /// Defaults to `.correction`: runs recorded before snippets existed have no such field,
    /// and failing to decode them would throw away the user's history to add a badge.
    public var kind: Kind

    public init(from: String, to: String, count: Int, kind: Kind = .correction) {
        self.from = from
        self.to = to
        self.count = count
        self.kind = kind
    }

    public init(from decoder: any Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        from = try container.decode(String.self, forKey: .from)
        to = try container.decode(String.self, forKey: .to)
        count = try container.decode(Int.self, forKey: .count)
        kind = try container.decodeIfPresent(Kind.self, forKey: .kind) ?? .correction
    }
}

/// Rewrites transcribed text using the dictionary's correction pairs.
///
/// This is the guaranteed half of the dictionary. Engine biasing is a nudge — it raises the
/// odds of the right word and promises nothing — so anything that must be correct has to be
/// fixed here, after the fact, deterministically.
///
/// Three rules, all load-bearing:
///
/// **Longest match first.** "Claude Code" is applied before "Claude", so the longer rule
/// isn't pre-empted by a shorter one that overlaps it.
///
/// **Whole matches only.** Every pattern is fenced by word boundaries, so a rule for
/// "cloud code" can never touch "Cloudflare" or the ordinary word "cloud".
///
/// **Glued words still match.** Engines run words together — "CloudCode", "cloud-code" — so
/// the gap between the parts of a phrase is matched as *optional* whitespace or hyphens
/// rather than a literal space.
public struct DictionaryCorrector: Sendable {
    private let corrections: [Rule]
    private let snippets: [Rule]
    /// What a correction-rule hit is reported as. Snippet hits always report `.snippet`.
    private let reportedKind: AppliedCorrection.Kind

    private struct Rule: Sendable {
        let regex: NSRegularExpression
        let replacement: String
        let trigger: String
    }

    public init(entries: [DictionaryEntry]) {
        corrections = Self.compile(entries, kind: .correction)
        snippets = Self.compile(entries, kind: .snippet)
        reportedKind = .correction
    }

    /// Compiles arbitrary "when you hear X, write Y" pairs that aren't dictionary entries.
    ///
    /// Exists so screen-context matching reuses this machinery rather than reimplementing
    /// it. Everything that makes the correction pass safe — NFC normalization, whole-match
    /// fences that can't bite into a longer word, tolerance for glued and hyphenated
    /// speech, longest-match-first ordering — is exactly what that feature needs, and a
    /// second copy of it is a second thing to get subtly wrong.
    ///
    /// - Parameter reportedAs: the kind recorded on every hit, so history can say where the
    ///   correction came from.
    public init(matching pairs: [(hear: String, write: String)], reportedAs kind: AppliedCorrection.Kind) {
        let entries = pairs.map { DictionaryEntry.correction(hear: $0.hear, write: $0.write) }
        corrections = Self.compile(entries, kind: .correction)
        snippets = []
        reportedKind = kind
    }

    /// Longest trigger first. Sorting by the trigger's length is what makes "Claude Code"
    /// win over "Claude" — once the longer rule has rewritten the span, the shorter one no
    /// longer sees the text it would have matched.
    private static func compile(_ entries: [DictionaryEntry], kind: DictionaryEntry.Kind) -> [Rule] {
        entries
            .filter { $0.isEnabled && $0.kind == kind }
            .filter { !$0.hear.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty }
            .sorted { $0.hear.count > $1.hear.count }
            .compactMap { entry in
                guard let regex = makeRegex(for: entry.hear) else { return nil }
                return Rule(
                    regex: regex,
                    replacement: NSRegularExpression.escapedTemplate(for: entry.write),
                    trigger: entry.hear
                )
            }
    }

    public var isEmpty: Bool { corrections.isEmpty && snippets.isEmpty }

    /// Applies every rule in order: all corrections, then all snippets.
    ///
    /// That order is part of the contract, and it falls out of what the two kinds are for.
    /// A correction repairs what the engine misheard, so it has to run before a trigger is
    /// matched — "insert my cignature" must become "insert my signature" before the snippet
    /// can fire. Running snippets second also means an expansion's body is never rewritten
    /// by a correction rule, which is what you want from text you typed out exactly as you
    /// meant it.
    ///
    /// - Returns: the rewritten text, plus one `AppliedCorrection` per rule that fired.
    public func apply(to text: String) -> (text: String, applied: [AppliedCorrection]) {
        guard !isEmpty, !text.isEmpty else { return (text, []) }

        // Normalize to NFC before matching. macOS hands back decomposed (NFC vs NFD) strings
        // in several places — a filesystem read of the dictionary being the obvious one — and
        // "café" decomposed is five scalars where composed is four. The pattern and the text
        // must be in the same form or an accented trigger silently never matches. The Windows
        // implementation normalizes identically; this is part of the shared contract.
        var result = text.precomposedStringWithCanonicalMapping
        var applied: [AppliedCorrection] = []

        for (rules, kind) in [(corrections, reportedKind), (snippets, .snippet)] {
            for rule in rules {
                let range = NSRange(result.startIndex..., in: result)
                let matches = rule.regex.numberOfMatches(in: result, range: range)
                guard matches > 0 else { continue }

                // Record what the engine actually produced, not the rule's trigger — seeing
                // the real mishearing is the point, and it can differ from the trigger in
                // case or spacing ("CloudCode" matched by "cloud code").
                let firstMatch = rule.regex.firstMatch(in: result, range: range)
                let heard = firstMatch
                    .flatMap { Range($0.range, in: result) }
                    .map { String(result[$0]) } ?? rule.trigger

                result = rule.regex.stringByReplacingMatches(
                    in: result,
                    range: range,
                    withTemplate: rule.replacement
                )

                applied.append(AppliedCorrection(
                    from: heard,
                    to: rule.replacement.replacingOccurrences(of: "\\", with: ""),
                    count: matches,
                    kind: kind
                ))
            }
        }

        return (result, applied)
    }

    /// Builds the pattern for one trigger phrase.
    ///
    /// The parts are joined with `[\s\-]*` — zero or more spaces or hyphens — which is what
    /// catches "CloudCode" and "Cloud-Code" alongside the spaced form.
    ///
    /// The fences are lookarounds on letters and digits rather than `\b`. `\b` would treat a
    /// trailing hyphen or apostrophe as a boundary and let a rule bite into a longer word;
    /// requiring that no letter or digit sits on either side is the stricter guarantee, and
    /// it's what keeps "cloud code" off "Cloudflare".
    private static func makeRegex(for trigger: String) -> NSRegularExpression? {
        // NFC here too, matching `apply(to:)` — a trigger typed into the UI and a trigger read
        // back from the dictionary file can arrive in different normal forms.
        let parts = trigger
            .precomposedStringWithCanonicalMapping
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .split(whereSeparator: { $0 == " " || $0 == "-" || $0 == "\t" })
            .map { NSRegularExpression.escapedPattern(for: String($0)) }

        guard !parts.isEmpty else { return nil }

        let body = parts.joined(separator: "[\\s\\-]*")
        let pattern = "(?<![\\p{L}\\p{N}])\(body)(?![\\p{L}\\p{N}])"

        return try? NSRegularExpression(pattern: pattern, options: [.caseInsensitive])
    }
}

// MARK: - Engine biasing

public extension DictionaryCorrector {
    /// The phrases to hand the speech engine as context before it transcribes.
    ///
    /// Kept deliberately short. These models drift when given a long context list — on quiet
    /// or ambiguous audio they start inventing text from the vocabulary they were primed
    /// with, which is a far worse failure than the misspelling it was meant to fix.
    public static let biasLimit = 40

    /// - Returns: the correct spellings — `.term` words and the *write* side of corrections —
    ///   most recently useful first, capped at `biasLimit`.
    public static func biasPhrases(from entries: [DictionaryEntry]) -> [String] {
        var seen = Set<String>()
        var phrases: [String] = []

        // Snippets are excluded deliberately. Their `write` side is a block of prose, and
        // the whole point of the cap below is that a long context list makes these models
        // hallucinate from their priming on quiet audio.
        for entry in entries where entry.isEnabled && entry.kind != .snippet {
            let phrase = entry.write.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !phrase.isEmpty, seen.insert(phrase.lowercased()).inserted else { continue }
            phrases.append(phrase)
            if phrases.count == biasLimit { break }
        }

        return phrases
    }
}

File: windows/src/Murmur.Dictionary/DictionaryCorrector.cs
using System.Text;
using System.Text.RegularExpressions;

namespace Murmur.Dictionary;

/// <summary>Which kind of rule fired, so history can tell a repair from an expansion.</summary>
public enum CorrectionKind
{
    /// <summary>A mishearing was repaired.</summary>
    Correction,

    /// <summary>A spoken trigger expanded into a block of text.</summary>
    Snippet,

    /// <summary>
    /// A word matched against something visible on screen — a filename, a symbol.
    /// </summary>
    /// <remarks>
    /// Produced by the macOS app's screen-context pass only; the shared correction pass
    /// never emits it. Present here so the two sides agree on the reporting vocabulary.
    /// </remarks>
    Screen,
}

/// <summary>One correction that actually fired.</summary>
/// <param name="From">The text as the engine produced it.</param>
/// <param name="To">What it was rewritten to.</param>
/// <param name="Count">How many times it fired in this transcript.</param>
/// <param name="Kind">Which kind of rule fired. Defaults to a correction.</param>
public sealed record AppliedCorrection(
    string From,
    string To,
    int Count,
    CorrectionKind Kind = CorrectionKind.Correction);

/// <summary>
/// Rewrites transcribed text using the dictionary's correction pairs.
/// </summary>
/// <remarks>
/// <para>
/// This is the guaranteed half of the dictionary. Engine biasing is a nudge — it raises the
/// odds of the right word and promises nothing — so anything that must be correct is fixed
/// here, after the fact, deterministically.
/// </para>
/// <para>
/// A direct counterpart to <c>DictionaryCorrector.swift</c>. The two are independent
/// implementations of one contract, and <c>shared/dictionary-test-vectors.json</c> is what
/// stops them drifting. <b>Change the semantics there first.</b>
/// </para>
/// <para>Three rules, all load-bearing:</para>
/// <list type="number">
/// <item><b>Longest match first.</b> "Claude Code" is applied before "Claude", so the longer
/// rule isn't pre-empted by a shorter one that overlaps it.</item>
/// <item><b>Whole matches only.</b> Every pattern is fenced by word boundaries, so a rule for
/// "cloud code" can never touch "Cloudflare" or the ordinary word "cloud".</item>
/// <item><b>Glued words still match.</b> Engines run words together — "CloudCode",
/// "cloud-code" — so the gap between parts is matched as optional whitespace or hyphens.</item>
/// </list>
/// </remarks>
public sealed class DictionaryCorrector
{
    /// <summary>
    /// Guards against a pathological dictionary hanging a dictation. Matching is linear here,
    /// so this should never trigger; it exists so that a bug can't wedge the app.
    /// </summary>
    private static readonly TimeSpan MatchTimeout = TimeSpan.FromSeconds(1);

    private static readonly char[] PhraseSeparators = [' ', '-', '\t'];

    private readonly List<Rule> _corrections;
    private readonly List<Rule> _snippets;

    /// <summary>What a correction-rule hit is reported as. Snippet hits report Snippet.</summary>
    private readonly CorrectionKind _reportedKind;

    private sealed record Rule(Regex Regex, string Replacement, string Trigger);

    /// <summary>Compiles the enabled entries into two ordered rule sets.</summary>
    /// <param name="entries">The dictionary. Terms and disabled entries are ignored here.</param>
    public DictionaryCorrector(IEnumerable<DictionaryEntry> entries)
    {
        var materialized = entries as IReadOnlyCollection<DictionaryEntry> ?? entries.ToList();
        _corrections = Compile(materialized, EntryKind.Correction);
        _snippets = Compile(materialized, EntryKind.Snippet);
        _reportedKind = CorrectionKind.Correction;
    }

    /// <summary>Compiles arbitrary "when you hear X, write Y" pairs that aren't entries.</summary>
    /// <param name="pairs">Trigger and replacement pairs.</param>
    /// <param name="reportedAs">The kind recorded on every hit.</param>
    /// <remarks>
    /// Mirrors the Swift initializer of the same shape, so screen-context matching reuses
    /// this machinery rather than reimplementing the fences and ordering that make the
    /// correction pass safe.
    /// </remarks>
    public DictionaryCorrector(
        IEnumerable<(string Hear, string Write)> pairs,
        CorrectionKind reportedAs)
    {
        var entries = pairs.Select(p => DictionaryEntry.Correction(p.Hear, p.Write)).ToList();
        _corrections = Compile(entries, EntryKind.Correction);
        _snippets = [];
        _reportedKind = reportedAs;
    }

    /// <summary>
    /// Longest trigger first. Sorting by the trigger's length is what makes "Claude Code" win
    /// over "Claude" — once the longer rule has rewritten the span, the shorter one no longer
    /// sees the text it would have matched.
    /// </summary>
    /// <remarks>
    /// OrderByDescending is a *stable* sort in LINQ, so equal-length triggers keep their file
    /// order on both platforms. Swift's sort is not stable, but ties can only occur between
    /// triggers of identical length, which cannot overlap the same span twice — so the
    /// observable result is the same either way.
    /// </remarks>
    private static List<Rule> Compile(IEnumerable<DictionaryEntry> entries, EntryKind kind) =>
        entries
            .Where(e => e.IsEnabled && e.Kind == kind)
            .Where(e => !string.IsNullOrWhiteSpace(e.Hear))
            .OrderByDescending(e => e.Hear.Length)
            .Select(e => MakeRule(e.Hear, e.Write))
            .OfType<Rule>()
            .ToList();

    /// <summary>True when no enabled entry produced a usable rule.</summary>
    public bool IsEmpty => _corrections.Count == 0 && _snippets.Count == 0;

    /// <summary>Applies every rule in order: all corrections, then all snippets.</summary>
    /// <param name="text">Raw transcribed text.</param>
    /// <returns>The rewritten text, plus one entry per rule that fired.</returns>
    /// <remarks>
    /// That order is part of the shared contract. A correction repairs what the engine
    /// misheard, so it has to run before a trigger is matched — "insert my cignature" must
    /// become "insert my signature" before the snippet can fire. Running snippets second also
    /// means an expansion's body is never rewritten by a correction rule, which is what you
    /// want from text the user typed out exactly as they meant it.
    /// </remarks>
    public (string Text, IReadOnlyList<AppliedCorrection> Applied) Apply(string text)
    {
        if (IsEmpty || string.IsNullOrEmpty(text)) return (text, []);

        // Normalize to NFC before matching, exactly as the Swift side does. Decomposed and
        // composed forms of the same accented word are different sequences of code points —
        // "café" is 4 or 5 depending on form — so an accented trigger silently never fires
        // unless both sides agree. This is part of the shared contract, not an optimisation.
        var result = text.Normalize(NormalizationForm.FormC);
        var applied = new List<AppliedCorrection>();

        foreach (var (rules, kind) in new[]
        {
            (_corrections, _reportedKind),
            (_snippets, CorrectionKind.Snippet),
        })
        {
            foreach (var rule in rules)
            {
                var matches = rule.Regex.Matches(result);
                if (matches.Count == 0) continue;

                // Record what the engine actually produced, not the rule's trigger — seeing
                // the real mishearing is the point, and it can differ in case or spacing
                // ("CloudCode" matched by "cloud code").
                var heard = matches[0].Value;

                // MatchEvaluator rather than a replacement string: it makes the replacement
                // strictly literal. A plain Replace would treat "$1", "$&" and friends in the
                // user's own text as substitutions, which is a real hazard when the
                // replacement is arbitrary user input.
                result = rule.Regex.Replace(result, _ => rule.Replacement);

                applied.Add(new AppliedCorrection(heard, rule.Replacement, matches.Count, kind));
            }
        }

        return (result, applied);
    }

    /// <summary>Builds the pattern for one trigger phrase.</summary>
    /// <remarks>
    /// <para>
    /// Parts are joined with <c>[\s\-]*</c> — zero or more spaces or hyphens — which catches
    /// "CloudCode" and "Cloud-Code" alongside the spaced form.
    /// </para>
    /// <para>
    /// The fences are lookarounds on letters and digits rather than <c>\b</c>. <c>\b</c>
    /// treats a trailing hyphen or apostrophe as a boundary and would let a rule bite into a
    /// longer word; requiring that no letter or digit sits on either side is the stricter
    /// guarantee, and it's what keeps "cloud code" off "Cloudflare".
    /// </para>
    /// </remarks>
    private static Rule? MakeRule(string trigger, string replacement)
    {
        // NFC here too, matching Apply(): a trigger typed into the UI and one read back from
        // the dictionary file can arrive in different normal forms.
        var parts = trigger
            .Normalize(NormalizationForm.FormC)
            .Trim()
            .Split(PhraseSeparators, StringSplitOptions.RemoveEmptyEntries)
            .Select(Regex.Escape)
            .ToArray();

        if (parts.Length == 0) return null;

        var body = string.Join(@"[\s\-]*", parts);
        var pattern = $@"(?<![\p{{L}}\p{{N}}]){body}(?![\p{{L}}\p{{N}}])";

        try
        {
            var regex = new Regex(pattern, RegexOptions.IgnoreCase | RegexOptions.CultureInvariant, MatchTimeout);
            return new Rule(regex, replacement, trigger);
        }
        catch (ArgumentException)
        {
            return null;
        }
    }

    // ---- Engine biasing ----

    /// <summary>
    /// How many phrases to hand the speech engine as context.
    /// </summary>
    /// <remarks>
    /// Deliberately small. These models drift when given a long context list — on quiet or
    /// ambiguous audio they start inventing text from the vocabulary they were primed with,
    /// which is a far worse failure than the misspelling it was meant to fix.
    /// </remarks>
    public const int BiasLimit = 40;

    /// <summary>
    /// The correct spellings — Term words and the <i>write</i> side of corrections, never a
    /// snippet body — capped at <see cref="BiasLimit"/>, de-duplicated case-insensitively,
    /// in file order.
    /// </summary>
    public static IReadOnlyList<string> BiasPhrases(IEnumerable<DictionaryEntry> entries)
    {
        var seen = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        var phrases = new List<string>();

        // Snippets are excluded deliberately. Their Write side is a block of prose, and the
        // whole point of the cap is that a long context list makes these models hallucinate
        // from their priming on quiet audio.
        foreach (var entry in entries.Where(e => e.IsEnabled && e.Kind != EntryKind.Snippet))
        {
            var phrase = entry.Write.Trim();
            if (phrase.Length == 0 || !seen.Add(phrase)) continue;
            phrases.Add(phrase);
            if (phrases.Count == BiasLimit) break;
        }

        return phrases;
    }
}
